plugins/org.eclipse.fordiac.ide.typeeditor/src/org/eclipse/fordiac/ide/typeeditor/AbstractTypeEditor.java

```diff
@@ -0,0 +1,458 @@
+		return commandStack;
+	}
+
+	@Override
+	public <T> T getAdapter(final Class<T> adapter) {
+		if (adapter == LibraryElement.class) {
+			return adapter.cast(getType());
+		}
+		if (adapter == CommandStack.class) {
+			return adapter.cast(getCommandStack());
+		}
+		if (adapter == IPropertySheetPage.class) {
+			return adapter.cast(new SelectionTabbedPropertySheetPage(this));
+		}
+		if (adapter == IGotoMarker.class) {
+			return adapter.cast(this);
+		}
+		if (adapter == GraphicalAnnotationModel.class) {
+			return adapter.cast(annotationModel);
+		}
+
+		if (isEditorActive()) {
+			// we should only call super if the editor is active otherwise we may get
+			// disposed errors
+			T result = super.getAdapter(adapter);
+			if (result == null && shouldCheckAllEditors(adapter)) {
+				result = getEditorPages().stream().map(innerEditor -> Adapters.adapt(innerEditor, adapter))
+						.filter(Objects::nonNull).findFirst().orElse(null);
+			}
+			return result;
+		}
+		return null;
+	}
+
+	@Override
+	public TypeEditorInput getEditorInput() {
+		return (TypeEditorInput) super.getEditorInput();
+	}
+
+	protected Collection<ITypeEditorPage> getEditorPages() {
+		return editorPages;
+	}
+
+	@Override
+	public String getTitleToolTip() {
+		final String tooltip = (getTypeEntry() != null) ? getTypeEntry().getFullTypeName() + "\n" : ""; //$NON-NLS-1$ //$NON-NLS-2$
+		return tooltip + super.getTitleToolTip();
+	}
+
+	protected LibraryElement getType() {
+		final TypeEditorInput ei = getEditorInput();
+		return (ei != null) ? ei.getContent() : null;
+	}
+
+	protected TypeEntry getTypeEntry() {
+		final TypeEditorInput ei = getEditorInput();
+		return (ei != null) ? ei.getTypeEntry() : null;
+	}
+
+	@Override
+	public void gotoMarker(final IMarker marker) {
+		int i = 0;
+		for (final ITypeEditorPage editorPart : getEditorPages()) {
+			if (editorPart.isMarkerTarget(marker)) {
+				setActivePage(i);
+				editorPart.gotoMarker(marker);
+				break;
+			}
+			i++;
+		}
+	}
+
+	public void handleContentOutlineSelection(final ISelection selection) {
+		if ((selection instanceof final IStructuredSelection sel) && !selection.isEmpty()) {
+			final Object selectedElement = sel.getFirstElement();
+			int i = 0;
+			for (final ITypeEditorPage editorPart : getEditorPages()) {
+				if (editorPart.outlineSelectionChanged(selectedElement)) {
+					if (getActivePage() != i) {
+						setActivePage(i);
+					}
+					break;
+				}
+				i++;
+			}
+		}
+	}
+
+	@Override
+	public void init(final IEditorSite site, final IEditorInput editorInput) throws PartInitException {
+		final TypeEditorInput typeEditorInput = checkEditorInput(editorInput);
+
+		if (isValidTypeEditorInput(typeEditorInput)) {
+			typeEditorInput.getTypeEntry().eAdapters().add(adapter);
+			annotationModel = new FordiacMarkerGraphicalAnnotationModel(typeEditorInput.getTypeEntry().getFile());
+			validationJob = new ValidationJob(getPartName(), getCommandStack(), annotationModel);
+		}
```

Comment:
I believe this should probably be moved to `setInput` below (also dispose old model and job, if present), in order to also handle later changes to the input?

Reply:
setInput is only called by us or by the TypeEntryAdapter in case of a new file. But I guess we need to do something in reload file, or?

Reply:
@mx990 added code that will update the file of the annotation model. I hope this meets the goals.